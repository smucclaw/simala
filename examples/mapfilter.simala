let opaque id =

    fun (x) => x in

let opaque cons =

    fun (x) => fun (xs) => x : xs in

let opaque map =
  
    fun (f, xs) =>
    foldr (fun (x, r) => f(x) : r, [], xs) in

let opaque filter =

    fun (p, xs) =>
    foldr (fun (x, r) => (if p(x) then cons(x) else id)(r), [], xs) in

  filter
    ( fun (x) => x % 2 == 0,
      map
        ( fun (x) => x * 3,
          [1,2,3,4,5,6,7,8,9,10]
        )
    )
