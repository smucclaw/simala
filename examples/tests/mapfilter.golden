id = <fun/1>
cons = <fun/1>
map = <fun/2>
filter = <fun/2>
[6,12,18,24,30]
